### ISSUE TITLE
Loading pretrain weights give bad result

### ISSUE THREAD

**Issue by Sunirban-Sarkar** [roboflow/rf-detr#258]
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

After training model saved two checkpoints files regular and ema. But when I again load those weights back for inference the results are bad or different. Can someone help?

### Environment

RF-DETR 1.1.0
OS Ubuntu 22.04 LTS
Python 3.10.18
Pytorch 2.7.1+cu128
Cuda 12.8
GPU RTX A5000

### Minimal Reproducible Example

Just load weight and try inference

### Are you willing to submit a PR?

- [ ] Yes, I'd like to help by submitting a PR!

**Comment by anyaflya** [comment]
# BUG REPORT: Detection head re-initialization

I stumbled into the same issue. I believe it's because of the `num_classes` mismatch, which leads to the re-initialization of the detection head. 

The `num_classes` mismatch issue was mentioned in https://github.com/roboflow/rf-detr/issues/218 and https://github.com/roboflow/rf-detr/issues/51. I believe it's a deep problem as it prevents prediction of a trained RF-DETR model. 

A PR https://github.com/roboflow/rf-detr/pull/261 was open, but still no update. 

### Environment

RF-DETR 1.2.1
OS Ubuntu 24.04.2 LTS
Python 3.12.7
Cuda 12.4
NVIDIA GeForce RTX 4080

### Minimal Reproducible Example
With a dataset with 2 classes, with ids starting at 0. 

#### training
```
model = RFDETRMedium()
model.train(dataset_dir=dataset_dir)
```
#### predict
```
model = RFDETRMedium(pretrain_weights="checkpoint_best_ema.pth", num_classes=2)
model.predict(image, threshold=0.5)
```

**Comment by SkalskiP** [comment]
Hi @Sunirban-Sarkar 👋🏻 Thanks so much for your interest in RF-DETR. I recommend using the latest version, `rfdetr==1.2.1`. I’ve just tested training the RF-DETR-Medium model on a custom dataset, benchmarking it, then loading the saved checkpoint from disk and benchmarking it again. In both cases, I got exactly the same results.

For more details, feel free to check out [this Google Colab](https://colab.research.google.com/drive/1d_jzhMkj7tbDUpjESSIUW2pHjN1zpp2Y?usp=sharing).

**Comment by Borda** [comment]
Hey @Sunirban-Sarkar, thanks for the report! Pretrained weight loading behaviour — including partial loads and the positional embedding mismatch that can affect accuracy — is tracked in #293.

Closing as a duplicate of #293. If you have more detail on what "bad results" looks like (e.g. mAP drop, random predictions), adding it to that thread would help build a clearer picture.
